### PR TITLE
Port Herblore plugin to Kotlin

### PR DESCRIPTION
**Pull request by ghost**
review :)

### CODE REVIEW

File: game/plugin/skills/herblore/src/potion.kt [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+class MakeFinishedPotionAction (val player: Player, val finished: FinishedPotion) : AsyncAction<Player>(1, false, player) {
+    override fun action(): ActionBlock = {
+        val level = mob.herblore.current
+
+        if (level < finished.level) {
+            player.sendMessage("You need an herblore level of ${finished.level} to make this potion.")
+            stop()
+        }
+
+        if (player.inventory.remove(finished.unf_potion, 1) > 0) {
```
Comment (garyttierney) [inline game/plugin/skills/herblore/src/potion.kt]: `player.inventory.remove(id, amount)` returns the number of items that was removed.  If you only want to remove one, you can write `player.inventory.remove(id)` intead, which returns a boolean indicating if one value was removed/.

File: game/plugin/skills/herblore/src/potion.kt [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    WEAPON_POISON(id = 187, unf_potion = UnfinishedPotion.KWUARM, ingredient = CrushIngredient.DRAGON_SCALE.crushed, level = 60, experience = 137.5),
+    SUPER_RESTORE(id = 3026, unf_potion = UnfinishedPotion.SNAPDRAGON, ingredient = NormalIngredient.RED_SPIDERS_EGGS.id, level = 63, experience = 142.5),
+    SUPER_DEFENCE(id = 163, unf_potion = UnfinishedPotion.CADANTINE, ingredient = NormalIngredient.WHITE_BERRIES.id, level = 66, experience = 150.0),
+    ANTIFIRE(id = 2428, unf_potion = UnfinishedPotion.LANTADYME, ingredient = CrushIngredient.DRAGON_SCALE.crushed, level = 69, experience = 157.5),
+    RANGING(id = 169, unf_potion = UnfinishedPotion.DWARF_WEED, ingredient = NormalIngredient.WINE_ZAMORAK.id, level = 72, experience = 162.5),
+    MAGIC(id = 3042, unf_potion = UnfinishedPotion.LANTADYME, ingredient = NormalIngredient.POTATO_CACTUS.id, level = 76, experience = 172.5),
+    ZAMORAK_BREW(id = 189, unf_potion = UnfinishedPotion.TORSTOL, ingredient = NormalIngredient.JANGERBERRIES.id, level = 78, experience = 175.0);
+
+    val unf_potion = unf_potion.id
```
Comment (garyttierney) [inline game/plugin/skills/herblore/src/potion.kt]: Only used internally.  May as well inline this where it's used.

File: game/plugin/skills/herblore/src/potion.kt [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    companion object {
+        private val FINISHED_POTION = FinishedPotion.values().associateBy  { Pair(it.unf_potion, it.ingredient) }
```
Comment (garyttierney) [inline game/plugin/skills/herblore/src/potion.kt]: i.e. 
```
Pair(it.unfinishedPotion.id, it.ingredient)
```

File: game/plugin/skills/herblore/src/potion.kt [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        other as MakeUnFinishedPotionAction
+        return mob == other.mob && unfinished == other.unfinished
+    }
+
+    override fun hashCode(): Int = Objects.hash(mob, unfinished)
+}
+
+enum class FinishedPotion (
+        val id: Int,
+        unf_potion: UnfinishedPotion,
```
Comment (garyttierney) [inline game/plugin/skills/herblore/src/potion.kt]: should be `camelCase`

File: game/plugin/skills/herblore/src/herblore.plugin.kts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        .where { option == 1 }
+        .then {
+            val unidentified = Herb[id] ?: return@then
+            it.startAction(IdentifyHerbAction(it, slot, unidentified))
+            terminate()
+        }
+
+on { ItemOnItemMessage::class }
+        .where { id == PESTLE_MORTAR || id == KNIFE }
+        .where { targetId == PESTLE_MORTAR || targetId == KNIFE }
```
Comment (garyttierney) [inline game/plugin/skills/herblore/src/herblore.plugin.kts]: You can only apply where once, so you'll need to fold this into a single expression.

```
.where { (id == PESTLE_MORTAR || id == KNIFE) && (targetId == PESTLE_MORTAR || targetId == KNIFE) }
```

CC @Major- thoughts?

File: game/plugin/skills/herblore/src/ingredient.kt [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    companion object {
+        private val CRUSH_INGREDIENT = CrushIngredient.values().associateBy { it.id }
+        operator fun get(id: Int) = CRUSH_INGREDIENT[id]
+    }
+}
+
+class CrushIngredientAction (val player: Player, val ingredient: CrushIngredient) : AsyncAction<Player>(1, false, player) {
+    override fun action(): ActionBlock = {
+        if (player.inventory.remove(ingredient.id, 1) > 0) {
```
Comment (garyttierney) [inline game/plugin/skills/herblore/src/ingredient.kt]: Same as other remove() calls below.

File: game/plugin/skills/herblore/src/potion.kt [hunk 7]
```diff
@@ -0,0 +1,131 @@
+import org.apollo.game.action.ActionBlock
+import org.apollo.game.action.AsyncAction
+import org.apollo.game.model.Animation
+import org.apollo.game.model.entity.Player
+import org.apollo.game.plugin.api.Definitions
+import org.apollo.game.plugin.api.herblore
+import java.util.*
+
+private val MIXING_ANIM = Animation(363)
+val WATER_VIAL = 227
```
Comment (garyttierney) [inline game/plugin/skills/herblore/src/potion.kt]: Is this used outside of here?  If not you can use `const val`

**Review by garyttierney — CHANGES_REQUESTED** [review]
Have some feedback but nothing major.  This looks pretty good.

**Comment by garyttierney** [comment]
cc @Major- we should be distributing code style settings at this point.

File: game/plugin/skills/herblore/src/herblore.plugin.kts (hunk 5 above)
Comment (Major-) [reply game/plugin/skills/herblore/src/herblore.plugin.kts]: We'll write some shim code to improve this as part of #395. Fine for now.

**Comment by garyttierney** [comment]
@chivvon there's also some code-style changes needed throughout - could you run your code through IntelliJ's auto formatter?  The .editorconfig we ship should apply most changes.

File: game/plugin/skills/herblore/src/potion.kt (hunk 2 above)
Comment (ghost) [reply game/plugin/skills/herblore/src/potion.kt]: What do you exactly mean with this? All the other things have been fixed now!

File: game/plugin/skills/herblore/src/potion.kt (hunk 3 above)
Comment (ghost) [reply game/plugin/skills/herblore/src/potion.kt]: What do you exactly mean with this? All the other things have been fixed now!

File: game/plugin/skills/herblore/src/potion.kt (hunk 2 above)
Comment (Major-) [reply game/plugin/skills/herblore/src/potion.kt]: Just means use the `UnfinishedPotion` as the value inside `FinishedPotion`, and call `unf_potion.id` (or `unfinished.id`, whatever it's called now). If you press `Ctrl+Shift+N` in IntelliJ it can do this for you.

**Comment by tlf30** [comment]
Looking good, thank you for the contribution!

File: game/plugin/skills/herblore/src/herblore.plugin.kts [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import org.apollo.game.message.impl.ItemOptionMessage
+import org.apollo.game.message.impl.ItemOnItemMessage
+
+on { ItemOptionMessage::class }
+        .where { option == 1 }
```
Comment (garyttierney) [inline game/plugin/skills/herblore/src/herblore.plugin.kts]: Could you reduce this to 4 spaces instead of 8?  I think .editorconfig should be doing that (cc @Major- if it's not).

File: game/plugin/skills/herblore/src/herblore.plugin.kts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+on { ItemOnItemMessage::class }
+        .where { (id == PESTLE_MORTAR || id == KNIFE) && (targetId == PESTLE_MORTAR || targetId == KNIFE) }
+        .then {
+            val raw = CrushIngredient[id] ?: return@then
+            it.startAction(CrushIngredientAction(it, raw))
+            terminate()
+        }
+
+on { ItemOnItemMessage::class }
+        . where { id == WATER_VIAL || targetId == WATER_VIAL }
```
Comment (garyttierney) [inline game/plugin/skills/herblore/src/herblore.plugin.kts]: Remove the extra whitespace before the dots here.

File: game/plugin/skills/herblore/src/herblore.plugin.kts (hunk 8 above)
Comment (garyttierney) [reply game/plugin/skills/herblore/src/herblore.plugin.kts]: (and same below)

File: game/plugin/skills/herblore/src/potion.kt [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+class MakeFinishedPotionAction(val player: Player, val finished: FinishedPotion) : AsyncAction<Player>(1, false, player) {
+    override fun action(): ActionBlock = {
+        val level = mob.herblore.current
+
+        if (level < finished.level) {
+            player.sendMessage("You need an herblore level of ${finished.level} to make this potion.")
+            stop()
+        }
+
+        if (player.inventory.remove(finished.unfinished.id)) {
```
Comment (garyttierney) [inline game/plugin/skills/herblore/src/potion.kt]: Ah, just re-read the `Inventory` docs. You can shorten this to: `player.inventory.remove(finished.unfinished.id, finished.ingredient)`.

File: game/plugin/skills/herblore/build.gradle [hunk 11]
```diff
@@ -0,0 +1,8 @@
+plugin {
+    name = "herblore_skill"
+    authors = [
+            "Chivvon"
```
Comment (garyttierney) [inline game/plugin/skills/herblore/build.gradle]: Same here with indentation (8 spaces to 4).

File: game/plugin/skills/herblore/src/potion.kt [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+    val herb = herb.identified
+
+    companion object {
+        private val UNFINISHED_POTION = UnfinishedPotion.values().associateBy { it -> it.herb }
+        operator fun get(id: Int) = UNFINISHED_POTION[id]
+    }
+}
+
+class MakeUnFinishedPotionAction(val player: Player, val unfinished: UnfinishedPotion) : AsyncAction<Player>(1, false, player) {
```
Comment (Major-) [inline game/plugin/skills/herblore/src/potion.kt]: `MakeUnfinishedPotionAction` please

File: game/plugin/skills/herblore/src/potion.kt [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+    override fun action(): ActionBlock = {
+        val level = mob.herblore.current
+
+        if (level < unfinished.level) {
+            player.sendMessage("You need an herblore level of ${unfinished.level} to make this unfinished potion.")
```
Comment (Major-) [inline game/plugin/skills/herblore/src/potion.kt]: should be "*a* Herblore level"

File: game/plugin/skills/herblore/src/potion.kt [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+            stop()
+        }
+
+        if (player.inventory.remove(WATER_VIAL)) {
+            if (player.inventory.remove(unfinished.herb)) {
+                player.inventory.add(unfinished.id)
+                player.sendMessage("You have successfully made the ${Definitions.item(unfinished.id)!!.name.toLowerCase()}.")
```
Comment (Major-) [inline game/plugin/skills/herblore/src/potion.kt]: This message should be `"You mix the ${Definitions.item(unfinished.id)!!.name.toLowerCase()}. into your potion."`

File: game/plugin/skills/herblore/src/potion.kt [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+enum class FinishedPotion(
+        val id: Int,
+        val unfinished: UnfinishedPotion,
+        val ingredient: Int,
+        val level: Int,
+        val experience: Double
+) {
+    ATTACK(id = 121, unfinished = UnfinishedPotion.GUAM, ingredient = NormalIngredient.EYE_NEWT.id, level = 1, experience = 25.0),
```
Comment (Major-) [inline game/plugin/skills/herblore/src/potion.kt]: Can you statically import all of the enumerators in `UnfinishedPotion`, `NormalIngredient`, and `CrushIngredient`

File: game/plugin/skills/herblore/src/ingredient.kt [hunk 16]
```diff
@@ -0,0 +1,68 @@
+import org.apollo.game.action.ActionBlock
+import org.apollo.game.action.AsyncAction
+import org.apollo.game.model.Animation
+import org.apollo.game.model.entity.Player
+import org.apollo.game.plugin.api.Definitions
+import java.util.*
+
+private val GRINDING_ANIM = Animation(364)
+val PESTLE_MORTAR = 233
```
Comment (Major-) [inline game/plugin/skills/herblore/src/ingredient.kt]: These ints can be `const`.

File: game/plugin/skills/herblore/src/ingredient.kt [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+val KNIFE = 5605
+
+enum class CrushIngredient (
+        val id: Int,
+        val crushed: Int
+) {
+    UNICORN_HORN(id = 237, crushed = 235),
```
Comment (Major-) [inline game/plugin/skills/herblore/src/ingredient.kt]: `CrushedIngredient` please

File: game/plugin/skills/herblore/src/identify.kt [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+        operator fun get(id: Int) = HERB[id]
+    }
+}
+
+class IdentifyHerbAction(val player: Player, val slot: Int, val herb: Herb) : AsyncAction<Player>(0, false, player) {
+    override fun action(): ActionBlock = {
+        val level = mob.herblore.current
+
+        if (level < herb.level) {
+            player.sendMessage("You need an herblore level of ${herb.level} to clean this herb.")
```
Comment (Major-) [inline game/plugin/skills/herblore/src/identify.kt]: Should be `a Herblore level`

File: game/plugin/skills/herblore/src/identify.kt [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+            stop()
+        }
+
+        if (player.inventory.removeSlot(slot, 1) > 0) {
+            player.inventory.add(herb.identified)
+            player.herblore.experience += herb.experience
+            player.sendMessage("You identify the herb as a ${Definitions.item(herb.identified)!!.name.toLowerCase()}.")
```
Comment (Major-) [inline game/plugin/skills/herblore/src/identify.kt]: Can you add a `// TODO handle indefinite article` comment here please

File: game/plugin/skills/herblore/src/herblore.plugin.kts [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+        . then {
+            val unfinished = UnfinishedPotion[id] ?: return@then
+            it.startAction(MakeUnFinishedPotionAction(it, unfinished))
+            terminate()
+        }
+
+on { ItemOnItemMessage::class }
+        . then {
+            val finished = FinishedPotion[Pair(targetId, id)] ?: return@then
```
Comment (Major-) [inline game/plugin/skills/herblore/src/herblore.plugin.kts]: You need to check which way round the `id` and `targetId` should go here - it needs to be `unfinished` first, `ingredient` second, so you should check which is which.

File: game/plugin/skills/herblore/src/herblore.plugin.kts [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+on { ItemOptionMessage::class }
+        .where { option == 1 }
+        .then {
+            val unidentified = Herb[id] ?: return@then
+            it.startAction(IdentifyHerbAction(it, slot, unidentified))
+            terminate()
+        }
+
+on { ItemOnItemMessage::class }
+        .where { (id == PESTLE_MORTAR || id == KNIFE) && (targetId == PESTLE_MORTAR || targetId == KNIFE) }
```
Comment (Major-) [inline game/plugin/skills/herblore/src/herblore.plugin.kts]: This check is wrong - will only pass if _both_ `id` and `targetId` are a knife or a pestle & mortar.

File: game/plugin/skills/herblore/src/herblore.plugin.kts [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+        .then {
+            val raw = CrushIngredient[id] ?: return@then
+            it.startAction(CrushIngredientAction(it, raw))
+            terminate()
+        }
+
+on { ItemOnItemMessage::class }
+        . where { id == WATER_VIAL || targetId == WATER_VIAL }
+        . then {
+            val unfinished = UnfinishedPotion[id] ?: return@then
```
Comment (Major-) [inline game/plugin/skills/herblore/src/herblore.plugin.kts]: Same thing with the `on` handler below, you need to make sure what you pass to `UnfinishedPotion` is actually the ingredient id (not the vial of water).

**Review by Major- — CHANGES_REQUESTED** [review]

File: game/plugin/skills/herblore/src/potion.kt (hunk 15 above)
Comment (ghost) [reply game/plugin/skills/herblore/src/potion.kt]: How would I do this if NormalIngredient and CrushedIngredient are using the val ingredient: Int?
Or did i just misunderstand what you ment?

Comment (garyttierney) [reply game/plugin/skills/herblore/src/potion.kt]: you can do:
```
import UnfinishedPotion.*
```
and change
```diff
- unfinished = UnfinishedPotion.GUAM
+ unfinished = GUAM
```

File: game/plugin/skills/herblore/src/herblore.plugin.kts (hunk 22 above)
Comment (ghost) [reply game/plugin/skills/herblore/src/herblore.plugin.kts]: That is the way I intially wanted to write the code. I just didn't know how to do it with plugin format you guys provided, "the on/where"-format. 

I'm still not sure how to do that logic in apollo with kotlin.

File: game/plugin/skills/herblore/src/potion.kt (hunk 15 above)
Comment (ghost) [reply game/plugin/skills/herblore/src/potion.kt]: Got it!

File: game/plugin/skills/herblore/src/herblore.plugin.kts (hunk 22 above)
Comment (Major-) [reply game/plugin/skills/herblore/src/herblore.plugin.kts]: Just something like:
```kotlin
val unfinishedId = if (id == WATER_VIAL) {
    targetId
} else {
    id
}

val unfinished = UnfinishedPotion[unfinishedId]!!
```

**Review by Major- — APPROVED** [review]

**Review by garyttierney — APPROVED** [review]
